src/sorting/merge_sort.rs

```diff
@@ -0,0 +1,157 @@
+/// Sort a slice using "merge sort" algorithm. `T` must impl `Ord`.
+///
+/// If `T` has zero sized, nothing happened.
+pub fn merge_sort<T>(slice: &mut [T])
+where
+    T: Ord,
+{
+    merge_sort_cmp(slice, |a, b| a.lt(b));
+}
+
+/// Sort a slice using "merge sort" algorithm. `less` is a `FnMut(&T, &T) -> bool`, `less(&lhs,
+/// &rhs)` means called defined that `lhs` is less than `rhs`. The result order is undefined if
+/// this `less` function not define a total ordering for elements in the slice.
+///
+/// If `T` has zero sized, nothing happened.
+pub fn merge_sort_cmp<T, F>(v: &mut [T], mut less: F)
+where
+    F: FnMut(&T, &T) -> bool,
```

Comment:
@plus7wist I love it, just one thing, do we really need the less function here?
We are requiring `T` to implement `Ord` trait, so can't we just use that?

Reply:
Thanks, for you love that code.

- Do we need the `less` function?

  Maybe not. This is just meaningful if we make `merge_sort_cmp` as a part of the result of our merge sort implemention. This is good for a practical library, but not necessary for a repo built for educational purpose. So I feel it ok to remove this `pub`, make `merge_sort` the only pubilc function.

- We are requiring `T` to implement `Ord` trait.

  The good thing is `merge_sort` need that, but `merge_sort_cmp` don't. We can use `merge_sort_cmp` to sort slice of type `T: !Ord`. And, we can sort same type with different order. like this sample:

  ```rust
  struct Point {
    x: i32,
    y: i32,
  }

  let points = /* something */;
  merge_sort_cmp(&mut points, |lhs, rhs| lhs.x < rhs.x);
  merge_sort_cmp(&mut points, |lhs, rhs| lhs.y < rhs.y);
  ```